### ISSUE TITLE
luksipc w/ lvm

### ISSUE THREAD

**Issue by mb-nrl** [johndoe31415/luksipc#1]
I was glad to see some solution to in-place encryption, but I also got confused how it works with the LVM or more specifically which device should be encrypted.

I have the following partitioning scheme:

/dev/sda1 ext2 boot
/dev/sda2 extended partition
/dev/sda3 LVM2 PV

This was a default scheme which installer applied when LVM option was chosen, but not encryption.  I thought that it might be possible to encrypt everything on /dev/sda2, basicaly convert it to a LUKS partition, but that cannot be done (0Mib = 0Gib chunks I believe and no fs).  Would converting /dev/sda3 be the way to go?  Or there is more to it?

What would the expected partitioning scheme be after running ipc with or without LVM?

**Comment by johndoe31415** [comment]
Hmmm, it really depends on what you want to do first and foremost. Basically what it boils down to is: do you want LVM on top of dm-crypt or do you want dm-crypt on top of LVM? Both is possible and both has its distinct advantages and disadvantages.

dm-crypt on top of LVM:
- For this you'd basically create a LUKS device on your logical volumes (e.g. /dev/my_bar_volgroup/my_foo_logvol)
- LVM doesn't need to know about this when you use luksipc, since you only deal with a file system on top of the logical volume's block device (just file system resizing as usual)
- You can selectively choose encryption for certain logical volume and not for others
- Also a disadvantage: If you want 2 of 3, for example, you'll have to do luksOpen twice every time you unlock your logical volumes

LVM on top of dm-crypt:
- For this, you'd need to encrypt your physical volume (PV)
- And LVM needs to be aware that you're about to do this, you need to resize your PV accordingly (this is very likely possible with LVM, but since I don't know LVM I can't tell you how it's done)
- All data that would is stored on this PV is automatically encrypted (every volume group, every logical volume)
- If you span partitions across multiple PVs and one is not encrypted, you'll have partially encrypted file systems on the lowest layer (on the upper layer they'll all be decrypted of course)

So, basically, that's your choice. I'd go for the first route (enrcypt logical volumes only, not physical volumes) because it appears to be much easier. But then again, I don't use LVM and am not an LVM expert.

Hope this helps,
Best regards,
Johannes

**Comment by mb-nrl** [comment]
Johannes,

Thanks for your thoughts.  I need LVM on top of dm-crypt (and that’s also the default for Ubuntu encryption) …  which makes things a bit more complicated.  I’ll keep looking for an answer.  Resizing PV was not an issue, but the tool does not work, I’m guessing because because there is no file system on the PV and the error said something about odd case with a very small drive (and the thing about 0Mib = 0Gib which was supposed to read /dev/X: 1024 MiB = 1.0 GiB).

Thanks.

— Magda

**Comment by johndoe31415** [comment]
Hey Magda,

hmmm this is very weird, because luksipc does not require a file system to work on. It works only on block device layer and doesn't even know about file systems.

Can you give the exact command line which failed and also the output ouf

# cat /proc/partitions

and

# sfdisk -l /dev/sda

(if sda is the correct device)? That would be very helpful so I can try to figure out what's going on for your.

Have you set your PV to inactive before trying to attempt the conversion?

Best regards,
Johannes

**Comment by mb-nrl** [comment]
Johannes,

I appreciate your help, but unfortunately, I decided to rebuild the machine (and turn on the encryption during installation) …  My other system got totally corrupted so I got desperate for a working one.

But  here is what I had …

It was a default install of Ubuntu with LVM option turned on during installation.

The partition table was 
sda1 /boot
sda2 efi
sda3 Extendend partition
sda5 LVM2 PV

There is a GPT  on /dev/sda so sfdisk was never happy.

The command that failed was

./luksipc -d /dev/sda3

I resized LV and PV before proceeding and unmounted everything, except the Live CD, before proceeding.  Maybe I should have been encrypting sda5 instead? 

The error I got said something about it being a rare case, small drive, and 0Mib = 0Gib.

I was encrypting sda3 because it seemed closer to the default encrypted Ubuntu configuration which ends up with sda3 as a LUKS partition, but clearly, my understanding of the relationship between those different types of partitions needs to be improved.

What is the intended target for luksinplace?  "For example, let's say you have a device at /dev/loop0 that has an ext4 file system.” made me think that it needed to be a specific type of partition so I assumed that that was the problem?  How does a /proc/partitions look like before or after for your /dev/loop0 example?

I have some more machines that need encrypting so I will give it a try again …  Maybe the rest don’t have LVM and things will be easier.  

Thanks.

— Magda

**Comment by johndoe31415** [comment]
Hi Magda,

extended partitions are only shown as pseudo-devices of which the size isn't the same extent as the contained logical partitions. Therefore, yes, the way you tried it must fail (because you selected the extended partition) and the one you would have needed to choose would have been /dev/sda5. luksipc detects that the partition is awkwardly short and bails out (I think it bails for every conversion process that is less than 128 MiB or something, because that's not a usecase for luksipc and most likely points to a user error).

The thing with loop devices is that they don't have a partition table. Therefore what you see in /proc/partitions is not what you'd see in a hard disk that has a paritition table with extended and logical partitions. But, as you described it, /dev/sda5 would have been the correct target.

If you want to check this in the future, you can issue

$ blockdev --getsz /dev/sda1
124997632

It will show you the size of the partition in sectors (512 bytes each). In my case this is a hard disk which is 124997632 blocks, i.e 124997632 * 512 / 1024 / 1024 / 1024 = 59.6 GiB in size. If the returned number seems weird you know something strange is going on.

Hope this clears it up a bit.

Best regards,
Johannes

**Comment by mb-nrl** [comment]
Great …   I’m queuing up another machine and hopefully it will go smoothly.  

Thanks so much for your help.  Have you tried talking to Ubuntu folks to make this tool part of the distribution? 

— Magda

**Comment by johndoe31415** [comment]
Sure, very happy to help out. But, if you're setting up a new machine, I'd do a cryptofs right from the start and not bother for luksipc. It's much easier and also better tested than luksipc :-)

No, haven't filed a ticket in Ubuntus Launchpad in order to get someone to package luksipc. I really don't know if it's important enough of a program to include in Ubuntu. After all, it's a rather narrow usecase and not without risks (even though I try my best to keep them as small as possible). So I'm not sure that Ubuntu would find it worthy for inclusion and I can't be bothered to try and convince people that it is.

If it really is good at doing its job (reliable & useful) then my guess is that someone will pick it up and package it. If it isn't (either unreliable or not useful) it won't be -- either way is fine with me.

I'll close this ticket if it's alright with you. If you have any more questions feel free to add some more.
Best regards,
Johannes

**Comment by mb-nrl** [comment]
Thanks!
